### ISSUE TITLE
Make the clean parameter in the file.directory state respect foreign require_in

### ISSUE THREAD

**Issue by micahhausler** [saltstack/salt#8646]
Right now the 'clean' parameter on the file.directory state will not remove files that are directly listed under 'require' or match the 'exclude_pat', but other files that have a 'require_in' referencing the directory will be deleted. The reason I don't have `my_app_config` explicitly in `my_directory.requires` is that I have multiple configs being generated dynamically. Can we make clean respect require_in?

Ex:
```
my_directory:
  file.directory:
    - name: /etc/my_app
    - clean: true

my_app_config:
  file.managed:
    - name: /etc/my_app/config.conf
    - require_in:
        - file: my_directory
```

`state.highstate` produces:

```
----------
    State: - file
    Name:      /etc/my_app/
    Function:  directory
        Result:    True
        Comment:   Files cleaned from directory /etc/my_app
        Changes:   removed: ['/etc/my_app/config.conf']
----------
    State: - file
    Name:      /etc/my_app/config.conf
    Function:  managed
        Result:    True
        Comment:   File /etc/config.conf updated
        Changes:   diff: New file

```

**Comment by FunTimeCoding** [comment]
+1

**Comment by bechtoldt** [comment]
+1

**Comment by multani** [comment]
So, I was facing the same issue the whole afternoon, except that I already had a state somewhere else which was working as expected (ie. ``require_in`` files don't get deleted) and I couldn't find why it was working in some cases and not with others.

I managed to find the difference. These states work as expected (directory gets cleared but files with ``require_in`` stay):
```
some_dir:
  file.directory:
    - name: /tmp/test
    - clean: true

/tmp/test/bar:
  file.managed:
    - name: /tmp/test/bar
    - require_in:
      - file: some_dir
```

These states trigger the bug reported by OP:
```
some_dir:
  file.directory:
    - name: /tmp/test
    - clean: true

some_file:
  file.managed:
    - name: /tmp/test/bar
    - require_in:
      - file: some_dir
```

So basically, it seems that if the ID of the file required in the directory is not the name of the to keep, then it gets ignored.

I haven't look at the code yet, but I will try to provide a patch for the next release.

**Comment by cachedout** [comment]
Closed via #27748
